controller_stopper/src/controller_stopper.cpp

```diff
@@ -0,0 +1,163 @@
+
+  if (stop_controllers_on_startup_ == true) {
+    while (stopped_controllers_.empty()) {
+      auto request = std::make_shared<controller_manager_msgs::srv::ListControllers::Request>();
+      auto future = controller_list_srv_->async_send_request(request);
+      rclcpp::spin_until_future_complete(node_, future);
+      auto result = future.get();
+      for (auto& controller : result->controller) {
+        // Check if in consistent_controllers
+        // Else:
+        //   Add to stopped_controllers
+        if (controller.state == "active") {
+          auto it = std::find(consistent_controllers_.begin(), consistent_controllers_.end(), controller.name);
+          if (it == consistent_controllers_.end() && controller.name.find("broadcaster") == std::string::npos) {
+            stopped_controllers_.push_back(controller.name);
+          }
+        }
+      }
+    }
+    auto request_switch_controller = std::make_shared<controller_manager_msgs::srv::SwitchController::Request>();
+    request_switch_controller->strictness = request_switch_controller->STRICT;
+    request_switch_controller->stop_controllers = stopped_controllers_;
+    auto future = controller_manager_srv_->async_send_request(request_switch_controller);
+    rclcpp::spin_until_future_complete(node_, future);
+    if (future.get()->ok == false) {
+      RCLCPP_ERROR(rclcpp::get_logger("Controller stopper"), "Could not deactivate requested controllers");
+    }
+    robot_running_ = false;
+  }
+}
+
+void ControllerStopper::findAndStopControllers()
+{
+  stopped_controllers_.clear();
+  auto request_switch_controller = std::make_shared<controller_manager_msgs::srv::SwitchController::Request>();
+  auto request_list_controllers = std::make_shared<controller_manager_msgs::srv::ListControllers::Request>();
+
+  // Callback to switch controllers
+  auto callback_switch_controller =
+      [this](rclcpp::Client<controller_manager_msgs::srv::SwitchController>::SharedFuture future_response) {
+        auto result = future_response.get();
+        if (result->ok == false) {
+          RCLCPP_ERROR(rclcpp::get_logger("Controller stopper"), "Could not deactivate requested controllers");
+        }
+      };
+
+  // Callback to list controllers
+  auto callback_list_controller =
+      [this, request_switch_controller, callback_switch_controller](
+          rclcpp::Client<controller_manager_msgs::srv::ListControllers>::SharedFuture future_response) {
+        auto result = future_response.get();
+        for (auto& controller : result->controller) {
+          // Check if in consistent_controllers
+          // Else:
+          //   Add to stopped_controllers
+          if (controller.state == "active") {
+            auto it = std::find(consistent_controllers_.begin(), consistent_controllers_.end(), controller.name);
+            if (it == consistent_controllers_.end() && controller.name.find("broadcaster") == std::string::npos) {
```

Comment:
I would rather not use such a string matching but rather rely on the consistent_controllers list solely.

```suggestion
            if (it == consistent_controllers_.end()) {
```